subsys/net/lib/ftp_client/src/ftp_client.c

```diff
@@ -0,0 +1,793 @@
+			sprintf(list_cmd, CMD_LIST_FILE, target);
+		} else {
+			memcpy(list_cmd, CMD_NLST, sizeof(CMD_NLST) - 1);
+		}
+	}
+	ret = do_ftp_send_ctrl(list_cmd, strlen(list_cmd));
+
+	/* Set up data connection */
+	if (client.work_q) {
+		k_work_submit_to_queue(client.work_q, &data_task_param.work);
+	} else {
+		k_work_submit(&data_task_param.work);
+	}
+
+	if (ret == 0) {
+		do {
+			ret = do_ftp_recv_ctrl(true, FTP_CODE_226);
+			if (ret == -ETIMEDOUT || ret == FTP_CODE_226) {
+				break;
+			}
+		} while (1);
+	}
+
+	return ret;
+}
+
+/**@brief API to change working directory
+ */
+int ftp_cwd(const char *folder)
+{
+	int ret;
+
+	if (strcmp(folder, "..") == 0) {
+		ret = do_ftp_send_ctrl(CMD_CDUP, sizeof(CMD_CDUP) - 1);
+	} else {
+		sprintf(ctrl_buf, CMD_CWD, folder);
+		ret = do_ftp_send_ctrl(ctrl_buf, strlen(ctrl_buf));
+	}
+	if (ret == 0) {
+		ret = do_ftp_recv_ctrl(true, FTP_CODE_250);
+	}
+
+	return ret;
+}
+
+/**@brief API to make directory
+ */
+int ftp_mkd(const char *folder)
+{
+	int ret;
+
+	sprintf(ctrl_buf, CMD_MKD, folder);
+	ret = do_ftp_send_ctrl(ctrl_buf, strlen(ctrl_buf));
+	if (ret == 0) {
+		ret = do_ftp_recv_ctrl(true, FTP_CODE_257);
+	}
+
+	return ret;
+}
+
+/**@brief API to remove directory
+ */
+int ftp_rmd(const char *folder)
+{
+	int ret;
+
+	sprintf(ctrl_buf, CMD_RMD, folder);
+	ret = do_ftp_send_ctrl(ctrl_buf, strlen(ctrl_buf));
+	if (ret == 0) {
+		ret = do_ftp_recv_ctrl(true, FTP_CODE_250);
+	}
+
+	return ret;
+}
+
+
+/**@brief API to rename file
+ */
+int ftp_rename(const char *old_name, const char *new_name)
+{
+	int ret;
+
+	sprintf(ctrl_buf, CMD_RNFR, old_name);
+	ret = do_ftp_send_ctrl(ctrl_buf, strlen(ctrl_buf));
+	if (ret == 0) {
+		ret = do_ftp_recv_ctrl(true, FTP_CODE_350);
+	}
+	if (ret != FTP_CODE_350) {
+		return ret;
+	}
+
+	sprintf(ctrl_buf, CMD_RNTO, new_name);
+	ret = do_ftp_send_ctrl(ctrl_buf, strlen(ctrl_buf));
+	if (ret == 0) {
+		ret = do_ftp_recv_ctrl(true, FTP_CODE_250);
+	}
+
+	return ret;
+}
+
+/**@brief API to delete a file
+ */
+int ftp_delete(const char *file)
+{
+	int ret;
+
+	sprintf(ctrl_buf, CMD_DELE, file);
+	ret = do_ftp_send_ctrl(ctrl_buf, strlen(ctrl_buf));
+	if (ret == 0) {
+		ret = do_ftp_recv_ctrl(true, FTP_CODE_250);
+	}
+
+	return ret;
+}
+
+/**@brief API to get a file
+ */
+int ftp_get(const char *file)
+{
+	int ret;
+	char get_cmd[128];
+
+	/* Always set Passive mode for nRF9160 to act as TCP client */
+	ret = do_ftp_send_ctrl(CMD_PASV, sizeof(CMD_PASV) - 1);
+	if (ret) {
+		return -EIO;
+	}
+	ret = do_ftp_recv_ctrl(true, FTP_CODE_227);
+	if (ret != FTP_CODE_227) {
+		return ret;
+	}
+	strcpy(data_task_param.ctrl_msg, ctrl_buf);
+	data_task_param.task = TASK_RECEIVE;
+
+	/* Send RETR command in control channel */
+	sprintf(get_cmd, CMD_RETR, file);
+	ret = do_ftp_send_ctrl(get_cmd, strlen(get_cmd));
+	if (ret) {
+		return -EIO;
+	}
+
+	/* Set up data connection */
+	k_work_submit_to_queue(client.work_q, &data_task_param.work);
+	k_sem_take(&rx_done, K_FOREVER);
```

Comment:
Actually, if we block in this function on a semaphore anyway, until the RX is done, why doesn't it call `do_ftp_recv_data` directly, instead of involving workqueue? Should be way simpler.

Reply:
Originally it's data channel and control channel receiving at the same time, thus data channel receive is started as k_work. Later tried to synchronized the output to AT response, by adding the semaphore. In reality the server send some control message, then data message, then control message. Sometimes the control message within data messages, too. The semaphore solution cannot cover all scenarios. It's not a problem for normal FTP client where control and data have different callbacks, but it matters for SLM, in which control and data are mixed up as AT response

Reply:
Ok, I think I understood the issue. Typically, you just `poll` on multiple sockets, and call `recv` on whatever socket indicated to have some data, but I guess mutli-threaded approach is also acceptable.